TestResourceManagerClientTestSuite is unstable

## Flaky Test

### Which jobs are failing
```
2023-07-12T10:45:54.8866587Z --- FAIL: TestResourceManagerClientTestSuite (28.56s)
2023-07-12T10:45:54.8867352Z     --- PASS: TestResourceManagerClientTestSuite/TestAcquireTokenBucket (4.24s)
2023-07-12T10:45:54.8868202Z     --- FAIL: TestResourceManagerClientTestSuite/TestBasicResourceGroupCURD (15.23s)
2023-07-12T10:45:54.8869052Z     --- FAIL: TestResourceManagerClientTestSuite/TestLoadRequestUnitConfig (0.01s)
2023-07-12T10:45:54.8869912Z     --- FAIL: TestResourceManagerClientTestSuite/TestRemoveStaleResourceGroup (0.00s)
2023-07-12T10:45:54.8870750Z     --- FAIL: TestResourceManagerClientTestSuite/TestResourceGroupController (0.00s)
2023-07-12T10:45:54.8871668Z     --- FAIL: TestResourceManagerClientTestSuite/TestResourceManagerClientDegradedMode (0.00s)
2023-07-12T10:45:54.8872886Z     --- FAIL: TestResourceManagerClientTestSuite/TestResourceManagerClientFailover (0.00s)
2023-07-12T10:45:54.8873715Z     --- FAIL: TestResourceManagerClientTestSuite/TestResourcePenalty (0.00s)
2023-07-12T10:45:54.8874594Z     --- FAIL: TestResourceManagerClientTestSuite/TestSwitchBurst (0.00s)
2023-07-12T10:45:54.8875389Z     --- FAIL: TestResourceManagerClientTestSuite/TestWatchResourceGroup (0.00s)
```
```
2023-07-12T10:45:54.8646141Z     resource_manager_test.go:972: 
2023-07-12T10:45:54.8646964Z         	Error Trace:	/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:972
2023-07-12T10:45:54.8647582Z         	Error:      	Received unexpected error:
2023-07-12T10:45:54.8648923Z         	            	[PD:client:ErrClientListResourceGroup]get all resource group failed, rpc error: code = Unavailable desc = not leader
2023-07-12T10:45:54.8649792Z         	Test:       	TestResourceManagerClientTestSuite/TestBasicResourceGroupCURD
2023-07-12T10:45:54.8650327Z     resource_manager_test.go:159: 
2023-07-12T10:45:54.8651163Z         	Error Trace:	/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:159
2023-07-12T10:45:54.8652595Z         	            				/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:153
2023-07-12T10:45:54.8653958Z         	            				/home/runner/go/pkg/mod/github.com/stretchr/testify@v1.8.2/suite/suite.go:179
2023-07-12T10:45:54.8655093Z         	            				/opt/hostedtoolcache/go/1.20.1/x64/src/runtime/panic.go:540
2023-07-12T10:45:54.8656224Z         	            				/opt/hostedtoolcache/go/1.20.1/x64/src/testing/testing.go:980
2023-07-12T10:45:54.8657620Z         	            				/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:972
2023-07-12T10:45:54.8658219Z         	Error:      	Received unexpected error:
2023-07-12T10:45:54.8659595Z         	            	[PD:client:ErrClientListResourceGroup]get all resource group failed, rpc error: code = Unavailable desc = not leader
2023-07-12T10:45:54.8660454Z         	Test:       	TestResourceManagerClientTestSuite/TestBasicResourceGroupCURD
2023-07-12T10:45:54.8661045Z === RUN   TestResourceManagerClientTestSuite/TestLoadRequestUnitConfig
2023-07-12T10:45:54.8662702Z [2023/07/12 10:45:50.975 +00:00] [INFO] [controller.go:138] ["load resource controller config"] [config="{\"degraded-mode-wait-duration\":\"1s\",\"request-unit\":{\"read-base-cost\":0.125,\"read-per-batch-base-cost\":0.5,\"read-cost-per-byte\":0.0000152587890625,\"write-base-cost\":1,\"write-per-batch-base-cost\":1,\"write-cost-per-byte\":0.0009765625,\"read-cpu-ms-cost\":0.3333333333333333}}"]
2023-07-12T10:45:54.8664712Z [2023/07/12 10:45:50.978 +00:00] [INFO] [controller.go:138] ["load resource controller config"] [config="{\"degraded-mode-wait-duration\":\"1s\",\"request-unit\":{\"read-base-cost\":1,\"read-per-batch-base-cost\":0,\"read-cost-per-byte\":2,\"write-base-cost\":3,\"write-per-batch-base-cost\":0,\"write-cost-per-byte\":4,\"read-cpu-ms-cost\":5}}"]
2023-07-12T10:45:54.8665667Z     resource_manager_test.go:159: 
2023-07-12T10:45:54.8666643Z         	Error Trace:	/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:159
2023-07-12T10:45:54.8668229Z         	            				/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:153
2023-07-12T10:45:54.8669569Z         	            				/home/runner/go/pkg/mod/github.com/stretchr/testify@v1.8.2/suite/suite.go:179
2023-07-12T10:45:54.8670826Z         	            				/home/runner/go/pkg/mod/github.com/stretchr/testify@v1.8.2/suite/suite.go:198
2023-07-12T10:45:54.8671383Z         	Error:      	Received unexpected error:
2023-07-12T10:45:54.8673614Z         	            	[PD:client:ErrClientListResourceGroup]get all resource group failed, rpc error: code = Unavailable desc = not leader
2023-07-12T10:45:54.8674524Z         	Test:       	TestResourceManagerClientTestSuite/TestLoadRequestUnitConfig
2023-07-12T10:45:54.8675133Z === RUN   TestResourceManagerClientTestSuite/TestRemoveStaleResourceGroup
2023-07-12T10:45:54.8675688Z     resource_manager_test.go:1116: 
2023-07-12T10:45:54.8676533Z         	Error Trace:	/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:1116
2023-07-12T10:45:54.8677144Z         	Error:      	Received unexpected error:
2023-07-12T10:45:54.8677901Z         	            	rpc error: code = Unavailable desc = not leader
2023-07-12T10:45:54.8678648Z         	Test:       	TestResourceManagerClientTestSuite/TestRemoveStaleResourceGroup
2023-07-12T10:45:54.8679182Z     resource_manager_test.go:159: 
2023-07-12T10:45:54.8680014Z         	Error Trace:	/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:159
2023-07-12T10:45:54.8681501Z         	            				/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:153
2023-07-12T10:45:54.8682808Z         	            				/home/runner/go/pkg/mod/github.com/stretchr/testify@v1.8.2/suite/suite.go:179
2023-07-12T10:45:54.8684101Z         	            				/opt/hostedtoolcache/go/1.20.1/x64/src/runtime/panic.go:540
2023-07-12T10:45:54.8685195Z         	            				/opt/hostedtoolcache/go/1.20.1/x64/src/testing/testing.go:980
2023-07-12T10:45:54.8686613Z         	            				/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:1116
2023-07-12T10:45:54.8687209Z         	Error:      	Received unexpected error:
2023-07-12T10:45:54.8688589Z         	            	[PD:client:ErrClientListResourceGroup]get all resource group failed, rpc error: code = Unavailable desc = not leader
2023-07-12T10:45:54.8689466Z         	Test:       	TestResourceManagerClientTestSuite/TestRemoveStaleResourceGroup
2023-07-12T10:45:54.8690077Z === RUN   TestResourceManagerClientTestSuite/TestResourceGroupController
2023-07-12T10:45:54.8690595Z     resource_manager_test.go:349: 
2023-07-12T10:45:54.8691408Z         	Error Trace:	/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:349
2023-07-12T10:45:54.8692016Z         	Error:      	Received unexpected error:
2023-07-12T10:45:54.8692806Z         	            	rpc error: code = Unavailable desc = not leader
2023-07-12T10:45:54.8693552Z         	Test:       	TestResourceManagerClientTestSuite/TestResourceGroupController
2023-07-12T10:45:54.8694076Z     resource_manager_test.go:159: 
2023-07-12T10:45:54.8694896Z         	Error Trace:	/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:159
2023-07-12T10:45:54.8696306Z         	            				/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:153
2023-07-12T10:45:54.8697619Z         	            				/home/runner/go/pkg/mod/github.com/stretchr/testify@v1.8.2/suite/suite.go:179
2023-07-12T10:45:54.8698761Z         	            				/opt/hostedtoolcache/go/1.20.1/x64/src/runtime/panic.go:540
2023-07-12T10:45:54.8699934Z         	            				/opt/hostedtoolcache/go/1.20.1/x64/src/testing/testing.go:980
2023-07-12T10:45:54.8701490Z         	            				/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:349
2023-07-12T10:45:54.8702078Z         	Error:      	Received unexpected error:
2023-07-12T10:45:54.8703581Z         	            	[PD:client:ErrClientListResourceGroup]get all resource group failed, rpc error: code = Unavailable desc = not leader
2023-07-12T10:45:54.8704442Z         	Test:       	TestResourceManagerClientTestSuite/TestResourceGroupController
2023-07-12T10:45:54.8705089Z === RUN   TestResourceManagerClientTestSuite/TestResourceManagerClientDegradedMode
2023-07-12T10:45:54.8705654Z     resource_manager_test.go:1032: 
2023-07-12T10:45:54.8706485Z         	Error Trace:	/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:1032
2023-07-12T10:45:54.8707095Z         	Error:      	Received unexpected error:
2023-07-12T10:45:54.8707889Z         	            	rpc error: code = Unavailable desc = not leader
2023-07-12T10:45:54.8708659Z         	Test:       	TestResourceManagerClientTestSuite/TestResourceManagerClientDegradedMode
2023-07-12T10:45:54.8709229Z     resource_manager_test.go:159: 
2023-07-12T10:45:54.8710054Z         	Error Trace:	/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:159
2023-07-12T10:45:54.8711506Z         	            				/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:153
2023-07-12T10:45:54.8712883Z         	            				/home/runner/go/pkg/mod/github.com/stretchr/testify@v1.8.2/suite/suite.go:179
2023-07-12T10:45:54.8713995Z         	            				/opt/hostedtoolcache/go/1.20.1/x64/src/runtime/panic.go:540
2023-07-12T10:45:54.8715128Z         	            				/opt/hostedtoolcache/go/1.20.1/x64/src/testing/testing.go:980
2023-07-12T10:45:54.8716507Z         	            				/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:1032
2023-07-12T10:45:54.8717094Z         	Error:      	Received unexpected error:
2023-07-12T10:45:54.8718488Z         	            	[PD:client:ErrClientListResourceGroup]get all resource group failed, rpc error: code = Unavailable desc = not leader
2023-07-12T10:45:54.8719422Z         	Test:       	TestResourceManagerClientTestSuite/TestResourceManagerClientDegradedMode
2023-07-12T10:45:54.8720087Z === RUN   TestResourceManagerClientTestSuite/TestResourceManagerClientFailover
2023-07-12T10:45:54.8720628Z     resource_manager_test.go:993: 
2023-07-12T10:45:54.8721428Z         	Error Trace:	/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:993
2023-07-12T10:45:54.8722033Z         	Error:      	Received unexpected error:
2023-07-12T10:45:54.8722825Z         	            	rpc error: code = Unavailable desc = not leader
2023-07-12T10:45:54.8723594Z         	Test:       	TestResourceManagerClientTestSuite/TestResourceManagerClientFailover
2023-07-12T10:45:54.8724144Z     resource_manager_test.go:159: 
2023-07-12T10:45:54.8724968Z         	Error Trace:	/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:159
2023-07-12T10:45:54.8726410Z         	            				/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:153
2023-07-12T10:45:54.8727720Z         	            				/home/runner/go/pkg/mod/github.com/stretchr/testify@v1.8.2/suite/suite.go:179
2023-07-12T10:45:54.8728838Z         	            				/opt/hostedtoolcache/go/1.20.1/x64/src/runtime/panic.go:540
2023-07-12T10:45:54.8729954Z         	            				/opt/hostedtoolcache/go/1.20.1/x64/src/testing/testing.go:980
2023-07-12T10:45:54.8731319Z         	            				/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:993
2023-07-12T10:45:54.8731906Z         	Error:      	Received unexpected error:
2023-07-12T10:45:54.8733280Z         	            	[PD:client:ErrClientListResourceGroup]get all resource group failed, rpc error: code = Unavailable desc = not leader
2023-07-12T10:45:54.8734163Z         	Test:       	TestResourceManagerClientTestSuite/TestResourceManagerClientFailover
2023-07-12T10:45:54.8734933Z === RUN   TestResourceManagerClientTestSuite/TestResourcePenalty
2023-07-12T10:45:54.8735436Z     resource_manager_test.go:559: 
2023-07-12T10:45:54.8736369Z         	Error Trace:	/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:559
2023-07-12T10:45:54.8736978Z         	Error:      	Received unexpected error:
2023-07-12T10:45:54.8737752Z         	            	rpc error: code = Unavailable desc = not leader
2023-07-12T10:45:54.8738438Z         	Test:       	TestResourceManagerClientTestSuite/TestResourcePenalty
2023-07-12T10:45:54.8738929Z     resource_manager_test.go:159: 
2023-07-12T10:45:54.8739741Z         	Error Trace:	/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:159
2023-07-12T10:45:54.8741220Z         	            				/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:153
2023-07-12T10:45:54.8742540Z         	            				/home/runner/go/pkg/mod/github.com/stretchr/testify@v1.8.2/suite/suite.go:179
2023-07-12T10:45:54.8743647Z         	            				/opt/hostedtoolcache/go/1.20.1/x64/src/runtime/panic.go:540
2023-07-12T10:45:54.8744779Z         	            				/opt/hostedtoolcache/go/1.20.1/x64/src/testing/testing.go:980
2023-07-12T10:45:54.8746137Z         	            				/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:559
2023-07-12T10:45:54.8746724Z         	Error:      	Received unexpected error:
2023-07-12T10:45:54.8748104Z         	            	[PD:client:ErrClientListResourceGroup]get all resource group failed, rpc error: code = Unavailable desc = not leader
2023-07-12T10:45:54.8748922Z         	Test:       	TestResourceManagerClientTestSuite/TestResourcePenalty
2023-07-12T10:45:54.8749462Z === RUN   TestResourceManagerClientTestSuite/TestSwitchBurst
2023-07-12T10:45:54.8749939Z     resource_manager_test.go:429: 
2023-07-12T10:45:54.8750759Z         	Error Trace:	/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:429
2023-07-12T10:45:54.8751358Z         	Error:      	Received unexpected error:
2023-07-12T10:45:54.8752241Z         	            	rpc error: code = Unavailable desc = not leader
2023-07-12T10:45:54.8752904Z         	Test:       	TestResourceManagerClientTestSuite/TestSwitchBurst
2023-07-12T10:45:54.8753391Z     resource_manager_test.go:159: 
2023-07-12T10:45:54.8754212Z         	Error Trace:	/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:159
2023-07-12T10:45:54.8755684Z         	            				/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:153
2023-07-12T10:45:54.8756957Z         	            				/home/runner/go/pkg/mod/github.com/stretchr/testify@v1.8.2/suite/suite.go:179
2023-07-12T10:45:54.8758081Z         	            				/opt/hostedtoolcache/go/1.20.1/x64/src/runtime/panic.go:540
2023-07-12T10:45:54.8759222Z         	            				/opt/hostedtoolcache/go/1.20.1/x64/src/testing/testing.go:980
2023-07-12T10:45:54.8760569Z         	            				/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:429
2023-07-12T10:45:54.8761171Z         	Error:      	Received unexpected error:
2023-07-12T10:45:54.8762559Z         	            	[PD:client:ErrClientListResourceGroup]get all resource group failed, rpc error: code = Unavailable desc = not leader
2023-07-12T10:45:54.8763361Z         	Test:       	TestResourceManagerClientTestSuite/TestSwitchBurst
2023-07-12T10:45:54.8763902Z === RUN   TestResourceManagerClientTestSuite/TestWatchResourceGroup
2023-07-12T10:45:54.8764407Z     resource_manager_test.go:198: 
2023-07-12T10:45:54.8765225Z         	Error Trace:	/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:198
2023-07-12T10:45:54.8765834Z         	Error:      	Received unexpected error:
2023-07-12T10:45:54.8766620Z         	            	rpc error: code = Unavailable desc = not leader
2023-07-12T10:45:54.8767472Z         	Test:       	TestResourceManagerClientTestSuite/TestWatchResourceGroup
2023-07-12T10:45:54.8767972Z     resource_manager_test.go:159: 
2023-07-12T10:45:54.8768908Z         	Error Trace:	/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:159
2023-07-12T10:45:54.8770395Z         	            				/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:153
2023-07-12T10:45:54.8771702Z         	            				/home/runner/go/pkg/mod/github.com/stretchr/testify@v1.8.2/suite/suite.go:179
2023-07-12T10:45:54.8772812Z         	            				/opt/hostedtoolcache/go/1.20.1/x64/src/runtime/panic.go:540
2023-07-12T10:45:54.8773920Z         	            				/opt/hostedtoolcache/go/1.20.1/x64/src/testing/testing.go:980
2023-07-12T10:45:54.8775253Z         	            				/home/runner/work/pd/pd/tests/integrations/mcs/resourcemanager/resource_manager_test.go:198
2023-07-12T10:45:54.8775846Z         	Error:      	Received unexpected error:
2023-07-12T10:45:54.8777225Z         	            	[PD:client:ErrClientListResourceGroup]get all resource group failed, rpc error: code = Unavailable desc = not leader
2023-07-12T10:45:54.8778070Z         	Test:       	TestResourceManagerClientTestSuite/TestWatchResourceGroup
```

### CI link
https://github.com/tikv/pd/actions/runs/5530616861/jobs/10090230566

### Reason for failure (if possible)


### Anything else

--- a/client/resource_manager_client.go
+++ b/client/resource_manager_client.go
@@ -36,7 +36,7 @@ const (
 	groupSettingsPathPrefix            = "resource_group/settings"
 	// errNotPrimary is returned when the requested server is not primary.
 	errNotPrimary = "not primary"
-	// errNotPrimary is returned when the requested server is not pd leader.
+	// errNotLeader is returned when the requested server is not pd leader.
 	errNotLeader = "not leader"
 )
 

--- a/client/resource_manager_client.go
+++ b/client/resource_manager_client.go
@@ -34,6 +34,8 @@ const (
 	add                     actionType = 0
 	modify                  actionType = 1
 	groupSettingsPathPrefix            = "resource_group/settings"
+	// errNotPrimary is returned when the requested server is not primary.
+	errNotPrimary = "not primary"
 	// errNotPrimary is returned when the requested server is not pd leader.
 	errNotLeader = "not leader"
 )
@@ -65,7 +67,7 @@ func (c *client) resourceManagerClient() (rmpb.ResourceManagerClient, error) {
 
 // gRPCErrorHandler is used to handle the gRPC error returned by the resource manager service.
 func (c *client) gRPCErrorHandler(err error) {
-	if strings.Contains(err.Error(), errNotLeader) {
+	if strings.Contains(err.Error(), errNotPrimary) || strings.Contains(err.Error(), errNotLeader) {
 		c.pdSvcDiscovery.ScheduleCheckMemberChanged()
 	}
 }

--- a/client/resource_manager_client.go
+++ b/client/resource_manager_client.go
@@ -34,8 +34,8 @@ const (
 	add                     actionType = 0
 	modify                  actionType = 1
 	groupSettingsPathPrefix            = "resource_group/settings"
-	// errNotPrimary is returned when the requested server is not primary.
-	errNotPrimary = "not primary"
+	// errNotPrimary is returned when the requested server is not pd leader.
+	errNotLeader = "not leader"
 )
 
 // GroupSettingsPathPrefixBytes is used to watch or get resource groups.
@@ -65,7 +65,7 @@ func (c *client) resourceManagerClient() (rmpb.ResourceManagerClient, error) {
 
 // gRPCErrorHandler is used to handle the gRPC error returned by the resource manager service.
 func (c *client) gRPCErrorHandler(err error) {
-	if strings.Contains(err.Error(), errNotPrimary) {
+	if strings.Contains(err.Error(), errNotLeader) {
 		c.pdSvcDiscovery.ScheduleCheckMemberChanged()
 	}
 }

--- a/tests/integrations/mcs/resourcemanager/resource_manager_test.go
+++ b/tests/integrations/mcs/resourcemanager/resource_manager_test.go
@@ -968,8 +968,12 @@ func (suite *resourceManagerClientTestSuite) TestBasicResourceGroupCURD() {
 	}
 	re.NoError(suite.cluster.RunServers(serverList))
 	suite.cluster.WaitLeader()
-	newGroups, err := cli.ListResourceGroups(suite.ctx)
-	re.NoError(err)
+	var newGroups []*rmpb.ResourceGroup
+	testutil.Eventually(suite.Require(), func() bool {
+		var err error
+		newGroups, err = cli.ListResourceGroups(suite.ctx)
+		return err == nil
+	}, testutil.WithWaitFor(time.Second))
 	re.Equal(groups, newGroups)
 }
 

